src/search/parser/lexical_analyzer.cc

```diff
@@ -0,0 +1,131 @@
+}
+
+static const vector<pair<TokenType, regex>> token_type_expressions = {
+    {TokenType::OPENING_PARENTHESIS, build_regex("\\(")},
+    {TokenType::CLOSING_PARENTHESIS, build_regex("\\)")},
+    {TokenType::OPENING_BRACKET, build_regex("\\[")},
+    {TokenType::CLOSING_BRACKET, build_regex("\\]")},
+    {TokenType::COMMA, build_regex(",")},
+    {TokenType::EQUALS, build_regex("=")},
+    {TokenType::FLOAT, build_regex(
+       R"([+-]?((\d*\.\d+|\d+\.)(e[+-]?\d+|[kmg])?)\b)",
+       regex_constants::ECMAScript | regex_constants::icase)},
+    {TokenType::INTEGER, build_regex(
+       R"([+-]?(infinity|\d+(e+?\d+|[kmg])?)\b)",
+       regex_constants::ECMAScript | regex_constants::icase)},
+    {TokenType::BOOLEAN, build_regex(R"(true|false)",
+       regex_constants::ECMAScript | regex_constants::icase)},
+    {TokenType::LET, build_regex("let",
+       regex_constants::ECMAScript | regex_constants::icase)},
+    {TokenType::IDENTIFIER, build_regex(R"([a-zA-Z_]\w*)")}
+};
+
+
+static vector<Token> split_tokens(const string &text) {
+    vector<Token> tokens;
+    auto start = text.begin();
+    auto end = text.end();
+    while (start != end) {
+        bool has_match = false;
+        smatch match;
+
+        for (const auto &type_and_expression : token_type_expressions) {
+            TokenType token_type = type_and_expression.first;
+            const regex &expression = type_and_expression.second;
+            if (regex_search(start, end, match,expression)) {
+                tokens.push_back({match[1], token_type});
+                start += match[0].length();
+                has_match = true;
+                break;
+            }
+        }
+        if (!has_match) {
+            // If the string text is longer than the terminal, then both, text
+            // and our error indicator will be broken into multiple lines. The
+            // error indicator is still at the right x coordinate, and the row
+            // in which the error occurs is indicated by the number of white
+            // lines between the 'text' and the indicator. Improvements are
+            // highly appreciated.
+            // I.e.
+            // aaaaaaaaaa@aaaaaaaaaaa
+            //           ^
+            // aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+            // aaaaaaaaaaaaaaaaaaaaaa@aaaaaaaaaaaaaaaaaaa
+            //
+            //                       ^
+            regex_search(start, end, match, build_regex("\\S+"));
+            assert(!match.empty());
+            cerr << "Unrecognized token: " << match[1] << endl << text << endl;
+            cerr << string(distance(text.begin(), start), ' ') << "^" << endl;
+            utils::exit_with(utils::ExitCode::SEARCH_INPUT_ERROR);
+        }
+    }
+    return tokens;
+}
+
+TokenStream::TokenStream(const std::string &text)
+    : tokens(split_tokens(text)), pos(0) {
+}
+
+bool TokenStream::has_tokens(int n) {
+    return utils::in_bounds(pos + n - 1, tokens);
```

Comment:
`has_tokens(0)` for pos=0 evaluates to false, although it should be true.  As this call is not useful at all. I am fine with forbidding it via an assertion. Would that mean that we do not need to fix the evaluation for has_tokens(0) as we see this as an invalid call anyway?

Reply:
Exactly, if we assert that it's greater than `0`, we don't need to fix the `in_bounds` test because it's clear that we're not supporting this value anyway. (And just to be clear, I don't think anyone would ever write `has_tokens(0)` literally because that's pointless, but some code written by someone down the line might do `has_tokens(k)` where `k` is a variable that might conceivably be 0. No need to support that, but good to be explicit with an assertion that we don't support that.)